src/dbnode/persist/fs/index_claims_manager.go

```diff
@@ -0,0 +1,101 @@
+package fs
+
+import (
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/retention"
+	"github.com/m3db/m3/src/x/clock"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+type indexClaimsManager struct {
+	sync.Mutex
+
+	filePathPrefix                string
+	nowFn                         clock.NowFn
+	nextIndexFileSetVolumeIndexFn nextIndexFileSetVolumeIndexFn
+
+	// Map of ns ID string -> blockStart -> volumeIndexClaim.
+	volumeIndexClaims map[string]map[xtime.UnixNano]volumeIndexClaim
+}
+
+type volumeIndexClaim struct {
+	volumeIndex int
+	blockStart  xtime.UnixNano
+}
+
+// NewIndexClaimsManager returns an instance of the index claim manager. This manages
+// concurrent claims for volume indices per ns and block start.
+// NB(bodu): There should be only a single shared index claim manager among all threads
```

Comment:
nit: does it make sense to make this method cache in instance and just return one if it already exists?

Reply:
Hmm so there are unit tests that rely on the CTOR returning separate instances of the index claims manager.

I could add a reset method to the index claims manager and explicitly reset before each test but that feels kinda awkward esp just for tests... 

I think it's better to just keep it as is for now. We have _some_ confidence we're using a single instance since we are setting it at the `storage.Options` level and propagating it through the db w/ validation checks if not set.